src/CommonLib/ThirdParty/PSOpenAD/Authentication.cs

```diff
@@ -0,0 +1,515 @@
+                        (context.Flags & InitiatorContextReturnFlags.ISC_RET_CONFIDENTIALITY) != 0;
+
+                    Span<Helpers.SecPkgContext_Sizes> sizes = stackalloc Helpers.SecPkgContext_Sizes[1];
+                    fixed (Helpers.SecPkgContext_Sizes* sizesPtr = sizes)
+                    {
+                        SSPI.QueryContextAttributes(_context, SecPkgAttribute.SECPKG_ATTR_SIZES,
+                            (IntPtr)sizesPtr);
+
+                        _trailerSize = sizes[0].cbSecurityTrailer;
+                        _blockSize = sizes[0].cbBlockSize;
+                    }
+                }
+
+                return context.OutputBuffers.Length > 0 ? context.OutputBuffers[0] : Array.Empty<byte>();
+            }
+        }
+    }
+
+    public override byte[] Wrap(ReadOnlySpan<byte> data, bool encrypt)
+    {
+        if (_context == null || !Complete)
+            throw new InvalidOperationException("Cannot wrap without a completed context");
+
+        unsafe
+        {
+            ArrayPool<byte> shared = ArrayPool<byte>.Shared;
+            byte[] token = shared.Rent((int)_trailerSize);
+            byte[] padding = shared.Rent((int)_blockSize);
+
+            try
+            {
+                fixed (byte* tokenPtr = token, dataPtr = data, paddingPtr = padding)
+                {
+                    Span<Helpers.SecBuffer> buffers = stackalloc Helpers.SecBuffer[3];
+                    buffers[0].BufferType = (UInt32)SecBufferType.SECBUFFER_TOKEN;
+                    buffers[0].cbBuffer = _trailerSize;
+                    buffers[0].pvBuffer = (IntPtr)tokenPtr;
+
+                    buffers[1].BufferType = (UInt32)SecBufferType.SECBUFFER_DATA;
+                    buffers[1].cbBuffer = (UInt32)data.Length;
+                    buffers[1].pvBuffer = (IntPtr)dataPtr;
+
+                    buffers[2].BufferType = (UInt32)SecBufferType.SECBUFFER_PADDING;
+                    buffers[2].cbBuffer = _blockSize;
+                    buffers[2].pvBuffer = (IntPtr)paddingPtr;
+
+                    UInt32 qop = encrypt ? 0 : 0x80000001; // SECQOP_WRAP_NO_ENCRYPT
+                    SSPI.EncryptMessage(_context, qop, buffers, NextSeqNo());
+
+                    byte[] wrapped = new byte[buffers[0].cbBuffer + buffers[1].cbBuffer + buffers[2].cbBuffer];
+                    int offset = 0;
+                    if (buffers[0].cbBuffer > 0)
+                    {
+                        Buffer.BlockCopy(token, 0, wrapped, offset, (int)buffers[0].cbBuffer);
+                        offset += (int)buffers[0].cbBuffer;
+                    }
+
+                    Marshal.Copy((IntPtr)dataPtr, wrapped, offset, (int)buffers[1].cbBuffer);
+                    offset += (int)buffers[1].cbBuffer;
+
+                    if (buffers[2].cbBuffer > 0)
+                    {
+                        Buffer.BlockCopy(padding, 0, wrapped, offset, (int)buffers[2].cbBuffer);
+                        offset += (int)buffers[2].cbBuffer;
```

Comment:
We can remove since we no longer use the `offset` after this if block, might have been kept for future changes, though?